### PR TITLE
Composite Eigenstrain tensors

### LINKED ISSUE #6205: Add CompositeEigenstrain
Using the same templated inheritance as `CompositeMobilityTensor` but with `RankTwoTensor` as the template argument.

### PR DESCRIPTION
**Pull request by dschwen**
This refactors the BASE CLASS that allows users to build arbitrary tensor types from a list of base tensors and derivative parsed materials. It is used to build
- Elasticity tensors (`RankFourTensor`, tensor mechanics)
- Mobility tensors (`RealValueTensor`, phase field)

And now 
- Eigenstrains (`RankTwoTensor`, tensor mechanics)

The Eigenstrain capability required a bit of refactoring, because the tensor itself is stored in a material property, but the derivatives are set as negative derivatives of the elastic strain. As this capability is used in both phase field _and_ tensor mechanics I've moved it into framework (with the end goal of removing the module interdependence between PF and TM). I'm planning to add a generic CompositeRankTwoTensor class as well.

This PR also adds a `GenericConstantRankTwoTensor` material that is needed to define the base tensors for the composite Eigenstrain.

Closes #6205

### CODE REVIEW

**Comment by dschwen** [comment]
@laagesen maybe you can review this?

File: framework/src/materials/GenericConstantRankTwoTensor.C [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+/****************************************************************/
+
+#include "GenericConstantRankTwoTensor.h"
+
+template<>
+InputParameters validParams<GenericConstantRankTwoTensor>()
+{
+  InputParameters params = validParams<Material>();
+  params.addRequiredParam<std::vector<Real> >("tensor_values", "Vector of values defining the constant rank two tensor");
+  params.addRequiredParam<MaterialPropertyName>("tensor_name", "Name of teh tensor material property to be created");
```
Comment (laagesen) [inline framework/src/materials/GenericConstantRankTwoTensor.C]: typo teh -> the

File: modules/tensor_mechanics/src/materials/CompositeEigenstrain.C [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+{
+  InputParameters params = CompositeTensorBase<RankTwoTensor, ComputeStressFreeStrainBase>::validParams();
+  params.addClassDescription("Assemble an Eigenstrain tensor from multiple tensor contributions weighted by material properties");
+  return params;
+}
+
+CompositeEigenstrain::CompositeEigenstrain(const InputParameters & parameters) :
+    CompositeTensorBase<RankTwoTensor, ComputeStressFreeStrainBase>(parameters)
+{
+  initializeDerivativeProperties(_base_name + "elastic_strain");
```
Comment (laagesen) [inline modules/tensor_mechanics/src/materials/CompositeEigenstrain.C]: Just wanted to make sure, did we settle on `elastic_strain` as the standard name? I seem to remember in some cases it was changed to `mechanical_strain` but I can't remember if it was changed back.

**Comment by laagesen** [comment]
Just added one question, otherwise looks good.

File: modules/tensor_mechanics/src/materials/CompositeEigenstrain.C (hunk 2 above)
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/CompositeEigenstrain.C]: elastic is correct here

**Comment by tophmatthews** [comment]
+1 this fix!

This caused some issues with our intel compiler. Do you guys test against an intel compiler?

**Pull request merged**